src/core/PluginPinConnector.cpp

```diff
@@ -0,0 +1,527 @@
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "PluginPinConnector.h"
+
+#include <QDomDocument>
+#include <QDomElement>
+#include <cassert>
+#include <stdexcept>
+
+#include "AudioEngine.h"
+#include "Model.h"
+#include "PluginPinConnectorView.h"
+
+namespace lmms
+{
+
+PluginPinConnector::PluginPinConnector(Model* parent)
+	: Model{parent}
+{
+	updateTrackChannels(s_totalTrackChannels);
+}
+
+PluginPinConnector::PluginPinConnector(int pluginInCount, int pluginOutCount, Model* parent)
+	: Model{parent}
+{
+	updateTrackChannels(s_totalTrackChannels);
+	setChannelCounts(pluginInCount, pluginOutCount);
+}
+
+auto PluginPinConnector::Matrix::channelName(int channel) const -> QString
+{
+	// Custom name (if supported)
+	assert(channel >= 0);
+	if (channel < static_cast<int>(channelNames.size()))
+	{
+		return channelNames[channel];
+	}
+
+	// A-Z
+	if (channel < 26)
+	{
+		return QChar::fromLatin1('A' + channel);
+	}
+
+	// AA-ZZ
+	channel -= 26;
+	if (channel < 26 * 26)
+	{
+		auto ret = QString{"AA"};
+		ret[0] = QChar::fromLatin1('A' + channel / 26);
+		ret[1] = QChar::fromLatin1('A' + channel % 26);
+		return ret;
+	}
+
+	throw std::invalid_argument{"Too many channels"};
+}
+
+void PluginPinConnector::setChannelCounts(int inCount, int outCount)
+{
+	if (m_trackChannelsUsed > MaxTrackChannels)
+	{
+		throw std::runtime_error{"Only up to 256 track channels are allowed"};
+	}
+
+	if (inCount < 0)
+	{
+		throw std::invalid_argument{"Invalid input count"};
+	}
+
+	if (outCount < 0)
+	{
+		throw std::invalid_argument{"Invalid output count"};
+	}
+
+	if (inCount == 0 && outCount == 0)
+	{
+		throw std::invalid_argument{"At least one port count must be non-zero"};
+	}
+
+	if (in().channelCount == inCount && out().channelCount == outCount)
+	{
+		// No action needed
+		return;
+	}
+
+	const bool initialSetup = in().channelCount == 0 && out().channelCount == 0;
+
+	setChannelCount(inCount, m_in);
+	setChannelCount(outCount, m_out);
+
+	if (initialSetup)
+	{
+		setDefaultConnections();
+	}
+
+	emit propertiesChanged();
+}
+
+void PluginPinConnector::setChannelCountIn(int inCount)
+{
+	setChannelCounts(inCount, out().channelCount);
+}
+
+void PluginPinConnector::setChannelCountOut(int outCount)
+{
+	setChannelCounts(in().channelCount, outCount);
+}
+
+void PluginPinConnector::setDefaultConnections()
+{
+	// Assumes pin maps are cleared
+	// TODO: Take into account channel groups?
+
+	assert(m_trackChannelsUsed >= 2);
+
+	auto setConnections = [](Matrix& matrix) {
+		switch (matrix.channelCount)
+		{
+			case 0: break;
+			case 1:
+				matrix.pins[0][0]->setValue(true);
+				matrix.pins[1][0]->setValue(true);
+				break;
+			default: // >= 2
+				matrix.pins[0][0]->setValue(true);
+				matrix.pins[1][1]->setValue(true);
+				break;
+		}
+	};
+
+	setConnections(m_in);
+	setConnections(m_out);
+}
+
+void PluginPinConnector::routeToPlugin(f_cnt_t frames, CoreAudioData in, SplitAudioData<sample_t> out)
```

Comment:
I'm not sure if it's worth it for function parameters that aren't references, though it looks like the method itself could be `const`.

Reply:
You could modify a parameter, even if not reference/pointer, inside the function - same as for local variables. In both cases, `const` assures to the reader that you won't do it. So I think it's worth the same.

In the end, I leave it to you, I just think it's not consistent to be so careful about "normal local" variables, but not about function parameter variables.

Reply:
According to the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-immutable),
> Function parameters passed by value are rarely mutated, but also rarely declared const. To avoid confusion and lots of false positives, don’t enforce this rule for function parameters.